spark8t/services.py

```diff
@@ -634,7 +634,7 @@ def __init__(
         self,
         kube_config_file: Union[str, Dict[str, Any]],
         context_name: Optional[str] = None,
-        kubectl_cmd: str = "kubectl",
+        kubectl_cmd: str = Defaults().kubectl_cmd,
```

Comment:
The shipped default wasn't possible to override within the whole execution chain.

Of course there can be other corners to take `Defaults` into account -- this is one solution.

Reply:
Uhm, I would probably place this within the cli submodule, and not as default in the class. I feel a bit uncomfortable that the defaults of the class depends on environment variables. 

I believe it is best to add this when instanciating the `registry` object in the scripts, e.g. [here](https://github.com/canonical/spark-k8s-toolkit-py/blob/DPE-2023_fixing_hardcoded_default/spark8t/cli/service_account_registry.py#L112) and [here](https://github.com/canonical/spark-k8s-toolkit-py/blob/DPE-2023_fixing_hardcoded_default/spark8t/cli/pyspark.py#L30), with something like

```
registry = K8sServiceAccountRegistry(..., kubectl_cmd = defaults.kubectl_cmd)
```

ps. By the way, we could factor out how we instanciate the registry object in `params.py`, similarly to what we are currently doing for `kube_interface`

Reply:
All right, in this case I'm proposing the following refactor + update.

First of all, I think it is a mistake to mix our CLI exectuables with library functions. Second, I think it's also a mistake to mix our parameter-parsing with additional logic. (As the latter results in 'growing' modules comibning mixed functionalities).

My proposal is the following:

```
spark8t/
    cli/
        <CLI executables>
    lib/
        <various lib funcitons, like>
        <our own parameter parsing function -- currently caled params.py>
        <interpreting our settings (CLI arguments, config files, env vars, hardcoded defaults, etc.) -- currently I caled it process_settings.py >
    resources/
        <k8s config>
    <other core modules>
```
Within this frame `process_settings.py` should hold stuff like `get_kube_interface()`, where convenient settings are applied.

I think this provides both a more organized structure, and the correct settings to be applied on the correct level.

> I feel a bit uncomfortable that the defaults of the class depends on environment variables.

The same module is using `Defaults` all across, "sticking" it on each `LightKube` or `SparkInterface` object that's instantiated.... So all those instantions depend on env vars.... (Sry, I may have got this comment wrong if it's not what you meant.)

Reply:
Uhm, I'm not sure about this. The idea of the `cli` submodule is to implement the command line interface, centralizing the business logic to accessing to the functionalities of the package via prompt command line. Ideally, one could expose the same functionalities through a REST api. In that case, beside `cli`, we could also have a submodule `rest` where we centralize argument parsing and structure of the endpoints (and of course all logic to parse url endpoints arguments should live within the `rest` module).  

Therefore, I'd really rather keep the business logic to parse the inputs from command line in the `cli` submodule, and not moving it into another more general `lib` submodule. 

> So all those instantions depend on env vars

Not really. I believe there is a subtle but relevant difference: they depend on the `Defaults` class, not on env vars. One could even create `Defaults` from configuration files or hard-coded dict (as one may need to do in tests).

In CLI, it makes sense to create a Default object from env vars. And this is indeed what we do [here](https://github.com/canonical/spark-k8s-toolkit-py/blob/main/spark8t/cli/__init__.py#L5). Using the REST api example as before, in that case, `Defaults` should probably be parsed from a config file, e.g.

```
defauls = Default(read_yaml('my-config-file.yaml`)
```

ps. Actually, [here](https://github.com/canonical/spark-k8s-toolkit-py/blob/main/spark8t/domain.py#L169) we provide the default to read this from the env, but I actually don't like this, and we should probably remove the default value of `environ` there as well. 

I hope this clarify the thoughts behind some of the choices.

Reply:
Splitting the response to 2 chapters, addressing both onging threads.

**Structure**:

I see your point regarding the `cli` submodule, and I get the very valid rationale behind. However, I'm missing a response on the original points that I've raised, and which, thus still reamain open. I believe that there should be a solution that satisfies both objectives. (As I guess "dumping" everything-alltogether to a `cli/` and separately in a `rest/` folder would not take us closer to a clean, organized structure either :-) )

Refreshing my original points: I beleive that it's a mistake to
 1. Mix executable "endpoint" scripts with helper/library modules (even if related).
   This is against a clear structure within the repository.
 2. Mix different functional aspects within the same module
   (In our case: CLI parameter parsing vs. initialization of core components of the application)
   
Addresssing both your points and mine, we could also go with this structure:
```
spark8t/
   cli/
      lib/
      bin/
```
or less clean/organized
```
spark8t/
   cli/
      lib/
      <exectuable1>
      <exectuable2>
      ...      
```
...but if I get it right we're trying to hold on to a mostly flat structure, which is contradicting to further embedded levels of directory hierarchy. 

Thus the best way to split executables from library modules --at least for the time being-- seemed to introduce a fairly traditional and conventional `lib/`. At least as long as no further fine-grained structure is required. In case additional functionalities may be added, we can still easily split `spark8t/lib/` into `spark8t/cli/lib`, `spark8t/rest/lib`. (Likely this wouldn't be the only structural change to be made if the functionality extends :-) ) 
  
I agree that a list of solutions could be good here, and we shouldn't forget about the point you've raised. I'm just emphasizing the importance of a clear structure both within the repository (1) and on the level of individual modules (2). I'm glad to find any solution complying to these objectives.

**Defaults:**

I'm not sure if I get your point. I assume that we have a fixed structure of our `Defaults`, as an interface schema. By clear specification of potential configuration options (regardless of what source the configuration comes from). Interface definition (typically like `Defaults().spark_home`, `Defaults.kubectl_cmt`) that we can safely refer to in all modules using `Defaults`.

In this context, I may be overlooking something (I'm not being sarcastic, I honestly mean it!!! :-) ), but on a 1st base I don't see the difference, or why the orignally proposed solution may depend more on ENV_VARs (the concern you've raised originally) than the following, already existing references within the same module (`LightKube` `class`):

https://github.com/canonical/spark-k8s-toolkit-py/blob/main/spark8t/services.py#L278
-> then the same interface deferred directly in `create()` -- where ALL these references may depend on `SPARK_HOME` (equally as the proposed solution may or may not depend on the `SPARK_KUBECTL` variable)
 - https://github.com/canonical/spark-k8s-toolkit-py/blob/main/spark8t/services.py#L525
 - https://github.com/canonical/spark-k8s-toolkit-py/blob/main/spark8t/services.py#L536
 - etc.
 
There are similar references all across the `services.py` module, also including the `SparkInterface` class.

I honestly mean that I may be missing a point, or design objective here, and that there may be an underlying difference that didn't come clear to me.
 
Of course I agree that for consistency, we should just rather have all `Defaults` passed to `KubeInterface` object initialization as well (as we do it for the equal-level sibling `LightKube` heir of `AbstractKubeInterface`)... Keeping the two implementations as close to each other as possible. I personally would rather vote for:

```
# We already have this
class LightKube(AbstractKubeInterface):
    def __init__(self, defaults: Defaults, ...):
        self.defaults = defaults
        [..]
    
    def create(...):
        with open(self.defaults.template_serviceaccount):
        [..]
     
# This could be adjusted
class KubeInterface(AbstractKubeInterface):
    class __init__(self, defaults: Defaults, ...):
        self.defaults = defaults
        [..]
    def create(...):
        # self.defaults.kubectl_cmd used here as needed
    
```